dissect/target/plugins/os/unix/linux/fortios/_os.py

```diff
@@ -528,6 +557,50 @@ def key_iv_for_kernel_hash(kernel_hash: str) -> AesKey | ChaCha20Key:
     raise ValueError(f"No known decryption keys for kernel hash: {kernel_hash}")
 
 
+def key_iv_from_keychain(target: Target, kernel_hash: str) -> list[ChaCha20Key | AesKey]:
+    """Return list of ChaCha20Key and AesKey from keychain.
+
+    When using the ``--keychain-file`` option, the CSV format is:
+
+        provider,key_type,identifier,value
+        ChaCha20Seed,passphrase,<kernel_hash>,<chacha20_seed>
+        ChaCha20Key,passphrase,<kernel_hash>,<chacha20_key>:<chacha20_iv>
+        AesKey,passphrase,<kernel_hash>,<aes_key>:<aes_iv>
+
+    When using the ``--keychain-value`` option, multiple keys are returned due to missing provider.
+
+    Args:
+        target: Target instance.
+        kernel_hash: SHA256 hash of the kernel file used to match key identifiers in the keychain.
+
+    Returns:
+        List of ChaCha20Key or AesKey.
+    """
+    keys: list[ChaCha20Key | AesKey] = []
+    keychain_keys = keychain.get_all_keys()
+
+    # We prioritize keys with matching identifier (kernel hash)
+    identifier_keys = [key for key in keychain_keys if key.identifier == kernel_hash]
+    other_keys = [key for key in keychain_keys if key.identifier is None]
+
+    for key in identifier_keys + other_keys:
+        if not isinstance(key.value, str):
+            continue
+        if key.provider in ("ChaCha20Seed", None) and len(key.value) == 64:
+            # 32 bytes hex string
+            key_data, key_iv = _kdf_7_4_x(key.value)
+            keys.append(ChaCha20Key(key_data, key_iv))
+        elif key.provider in ("AesKey", None) and len(key.value) == 97:
+            # 48 bytes hex string with colon separator
+            key_data, _, key_iv = key.value.partition(":")
+            keys.append(AesKey(key_data, key_iv))
+        elif key.provider in ("ChaCha20Key", None) and len(key.value) == 97:
```

Comment:
The provider is currently a bit more slug-ish and descriptive to what it's for (`bde` for BitLocker, `luks` for LUKS). Maybe put `fortios` or `fortios-chacha20seed` or something like that?